Improve mason external --setup

ankingcodes opened the pull request:
This PR makes the following improvements : 
- split `spack` installation between `spack` command line tool (pinned version) and `spack-registry` (always pulls master branch).
- implemented "shallow clone" to install `spack` and `spack-registry`, reducing time of download to under a minute.
- implemented function to update from earlier version of `spack` and mason to latest version.
- modified `repos.yaml` file to point to `spack-registry`
- added appropriate error messages.
- updated the documentation.


fixes #15616

ankingcodes commented:
Requesting a review from @ben-albrecht. 
Please let me know for any changes 👍

ankingcodes commented:
> A) Generate a site-specific repos.yaml in $(prefix)/etc/spack/, where $(prefix) is $MASON_HOME/spack.

@ben-albrecht  Let's make sure I understand this properly. So `mason external --setup` clones `spack` and `spack-registry`, but our mason external commands use `spack`'s package repository, so we need to link the package repository of `spack` to `spack-registry` by modifying the `repos.yaml` inside of `$(prefix)/etc/spack/`. Right ? 

Thinking otherwise, what if we don't clone `spack` and `spack-registry` separately ? I believe an updated version of spack (branch releases/v0.14) should have the latest and working version of the spack CLI commands. (Correct me if I am wrong). Why not consider using the latest version of spack and update our mason external commands to work correctly with them, if they dont ?
This would also reduce time for installation and upgrading of mason problem for users using previous version of Mason.

ankingcodes commented:
I tested out some `mason external` commands with only a single clone of `spack`'s latest version, i.e, branch `releases/v0.14`. Most of them worked fine. If we agree to my suggestion, the only thing left to do would be points 2. and 3. of what you suggested above in upgrading Mason.

ben-albrecht commented:
> Why not consider using the latest version of spack and update our mason external commands to work correctly with them, if they dont ?
This would also reduce time for installation and upgrading of mason problem for users using previous version of Mason.

The issue with this is that we will want `mason external --setup` to always pull the master branch of the spack repository, but keep the spack CLI tool we use pinned to a version that is known to work. In the future we may want to add `mason external --update` or something to update the spack repository to the latest master branch.

ankingcodes commented:
> If we used the same master branch for both, then spack (which is still pre-1.0) could introduce a breaking change to their CLI on master

Okay. I'm convinced by this.

ankingcodes commented:
@ben-albrecht Added feature for generating site specific `repos.yaml` and added error messages related to "upgrade Mason". (specifically 1 and 4). 
Some doubts : 
> mason external --setup checks if the spack version matches the pinned version (only if SPACK_ROOT is not set) and checks out the new version if not.
 
  - Which version do you refer to here? `spack` or `spack-registry` ?
  - What does "checks out the new version" mean here ?

ben-albrecht commented:
>> mason external --setup checks if the spack version matches the pinned version (only if SPACK_ROOT is not set) and checks out the new version if not.

> Which version do you refer to here? spack or spack-registry ?

`mason external --setup` should check if the spack command line tool (`spack --version`) matches the version we have pinned in mason (`config const version = 'releases/v0.14';`)

> What does "checks out the new version" mean here ?

If it doesn't match the version we require, we need to update the version. This could be done through a `git fetch` and maybe something like `git checkout tags/{version} -b {version}-branch`. We could alternatively blow away our spack install and do a fresh `git clone --depth=1 -q branches/{version} ...`, but I would only revert to this if we ran into difficulties with the former approach.

ankingcodes commented:
@ben-albrecht 
Added the following features : -
- command `mason external -V` to check `spack version` displayed as `0.14.2`
- implemented 2 from [comment](https://github.com/chapel-lang/chapel/pull/15641#pullrequestreview-406711040)
- implemented 3 from [comment](https://github.com/chapel-lang/chapel/pull/15641#pullrequestreview-406711040) 

Please take a look at the commands at fe0c83d. 
I have added `-C` to perform git operations inside `MASON_HOME` 
It works perfectly to update the version of spack, but when I then run other commands, I get the following printed, for example for checking version : 
```
$ mason external -V

fatal: No tags can describe '7a68a4d8513cd6105ee21bedfe04664c292d1e7d'. 
Try --always, or create some tags.
0.14.2 
fatal: No tags can describe '7a68a4d8513cd6105ee21bedfe04664c292d1e7d'. 
Try --always, or create some tags.
```
Perhaps it's due to using `-C`. Please share if you have other alternatives. Other than this, I performed some basic tests and everything seems to work fine.

ben-albrecht commented:
Any idea where `7a68a4d8513cd6105ee21bedfe04664c292d1e7d` is coming from? Is this maybe related to the fact that checking out the tag creates a detached head?

ankingcodes commented:
@ben-albrecht Actually we are not checking out a tag. We fetch the branch `releases/v0.14.2` to a `FETCH_HEAD` and then we checkout to a new branch `{version}-branch` as a commit, so that's where the SHA comes from.

ankingcodes commented:
@ben-albrecht This is ready for a review. Let me know for changes. 👍

ankingcodes commented:
> Something I thought of while writing this out -- mason external --setup should not do anything except print a friendly error message when SPACK_ROOT is set. I'm not sure that is the current behavior today..

We get a message "Spack backend is already installed" on running `mason external --setup` when SPACK_ROOT is set.

ben-albrecht commented:
> > Something I thought of while writing this out -- mason external --setup should not do anything except print a friendly error message when SPACK_ROOT is set. I'm not sure that is the current behavior today..
> 
> We get a message "Spack backend is already installed" on running `mason external --setup` when SPACK_ROOT is set.
> 

Excellent. Thanks for checking that.

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -18,7 +18,9 @@
 
-config const spackVersion = "releases/v0.11.2";
+config const version = '0.14.2';
+config const branch = 'releases/v0.14';
```
It'd be ideal to construct this string from `spackVersion` if that's safe to do, i.e. prepend `releases/v` and remove the last 2 characters. The advantage being that it is one less thing to update each time we update.

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -41,7 +43,24 @@ proc masonExternal(args: [] string) {
     }
     else if args[2] == "--setup" {
-      if isDir(SPACK_ROOT) then
+      if !isDir(getSpackRegistry) && isDir(SPACK_ROOT) then {
```
I'll only mention this once, but you can apply elsewhere: `then` is only needed when omitting the braces, e.g.

```chpl
if true then writeln('Ankush is a great student!');
```

```suggestion
      if !isDir(getSpackRegistry) && isDir(SPACK_ROOT)  {
```

ben-albrecht replied:
This condition probably deserves some explanation in a comment

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -41,7 +43,24 @@ proc masonExternal(args: [] string) {
+          throw new owned MasonError("Spack Registry installation failed");
+        } else exit(0);
+      }
+      else if isDir(SPACK_ROOT) &&
```
These conditions might benefit from a comment too.

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -83,22 +103,90 @@ private proc spackInstalled() throws {
+  else return 0;
+}
+
+proc cloneSpackRegistry() {
```
I'd advocate for adding a `/* doc-string */` to every function, even if very simple due to the procedure name being "self-documenting".

ben-albrecht commented:
The rst (online) documentation should also be updated to mention that `mason external --setup` always pulls from the master branch for the spack registry.

ankingcodes commented:
@ben-albrecht Done ✔️

ankingcodes commented on tools/mason/MasonHelp.chpl:
```diff
@@ -193,6 +193,7 @@ proc masonExternalHelp() {
   writeln("    uninstall                   Uninstall an external package");
   writeln("    info                        Show information about an external package");
   writeln("    find                        Find information about installed external packages");
+  writeln("    -V                          Display Spack version");
```
@ben-albrecht I did try to add a `--version` before, but on running it I got an error as follows: 
```
<command-line arg>:2: error: Configuration variable 'version' is missing its initialization value 
```

ben-albrecht replied:
Oh! That's because you define `config const version` above. Let's just make that a `const version` instead. (`const spackVersion` with my name suggestion)

ankingcodes commented on tools/mason/MasonExternal.chpl:
```diff
@@ -83,22 +103,90 @@ private proc spackInstalled() throws {
+  if isFile(yamlFilePath) then {
+    remove(yamlFilePath);
   }
+  const reposOverride = 'repos:\n'+
```
There's a linting issue with using `"""` for VSCode. I think other developers had raised an issue about this. Is it okay if I stick to using "+" for now?

ben-albrecht replied:
Yeah, not an issue

ankingcodes commented:
@ben-albrecht Made changes suggested. Ready for testing.

ben-albrecht commented on doc/rst/tools/mason/mason.rst:
```diff
@@ -548,7 +548,8 @@ First, the Spack backend must be installed. Users can have mason install Spack
 or point mason to an existing spack installation.
 
 This command will install Spack into ``$MASON_HOME`` and set it up so that it
```
```suggestion
This command will install Spack into ``$MASON_HOME/spack`` and set it up so that it
```

ben-albrecht commented on doc/rst/tools/mason/mason.rst:
```diff
@@ -548,7 +548,8 @@ First, the Spack backend must be installed. Users can have mason install Spack
 This command will install Spack into ``$MASON_HOME`` and set it up so that it
-can be used by Mason::
+can be used by Mason. It should be noted that this command pulls from the `master` branch of spack
+for setting up the spack registry at ``$MASON_HOME``::
```
```suggestion
for setting up the spack registry at ``$MASON_HOME/spack-registry``::
```

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -18,7 +18,12 @@
  */
 
-config const spackVersion = "releases/v0.11.2";
+const spackVersion = '0.14.2';
```
Let's put the Chapel version number in here to remind us to update this next release:

```suggestion
/* Version as of Chapel 1.22 - to be updated each release */
const spackVersion = '0.14.2';
```

ankingcodes commented:
@ben-albrecht In the test where you had to run `mason external --setup` twice. The error message was displayed twice because :
- the `spack-registry` was absent.
- the version of `spack` didnt match to latest version.
I think we should improve the error message to fix this, because our logic is already very complex. 
What do u think?

ben-albrecht commented:
> @ben-albrecht In the test where you had to run `mason external --setup` twice. The error message was displayed twice because :
> 
>     * the `spack-registry` was absent.
> 
>     * the version of `spack` didnt match to latest version.
> 


I am not sure I understand. Shouldn't the first `mason external --setup` have created the `spack-registry` and updated the `spack` version?

ankingcodes commented:
@ben-albrecht  We had different checks for them. There are 3 checks : -
- if `spack` is not present, get latest version of `spack` 
- if `spack-registry` is not present, get latest version of `spack-registry`
- if both `spack` and `spack registry` are present, but version is not latest, then update version.

ben-albrecht commented:
I see, so maybe the logic should go something like:

- if spack is not present OR if spack is not the latest version, get latest version of spack
- if spack-registry is not present, get latest version of spack-registry

and execute both of these checks `mason external --setup`. Does that make sense?

ben-albrecht commented:
> Another solution would be to refactor the checking function completely because the logic is already getting complex.

I think I'm suggesting this route 😄

ankingcodes commented:
@ben-albrecht made changes suggested. Regarding the `mason build` command, AFAIK, it interacts with the spack commands. I am unsure where we should place our error message. We did modify the `repos.yaml` file, I hope that would be good enough ?

ben-albrecht commented:
>> Until the user updates, will mason build still look in the right place for
externally installed packages, or will it try to look in $MASON_HOME/spack-registry which may not exist yet? 

> We did modify the repos.yaml file inside spack and pointed it to spack-registry, I hope that would be good enough ?

Oh yeah, that's right. That seems like the ideal behavior.  👍 

> it should probably print a warning that the user's spack setup is outdated and needs to run mason external --setup

It would still be great to emit a warning if the user is using external (spack) packages and has not updated. Can we add a check to `MasonBuild.chpl:buildProgram()` that does something like:

```chpl
var sourceList = genSourceList(lockFile);
if lockFile.pathExists('external') {
  if MasonExternal.spackOutdated() || MasonExternal.spackRegistryOutdated() {
    writeln('<our warning message about updating mason external>');
  }
}
```

This will require use to refactor the checks in `MasonExternal` into functions, which should make those checks a bit cleaner as well.

ankingcodes commented:
@ben-albrecht I hope this covers all cases now.

ankingcodes commented:
@ben-albrecht I've added the following error message for the above mentioned error : 
```
Mason could not find Spack at $MASON_HOME. To install Spack, use `mason external --setup`
```

ben-albrecht commented:
@ankingcodes - I'm still seeing the fatal tag exists issue on the first time "upgrading" the backend:

```sh
[master branch]$ mason external --setup
[ankingcodes:setup]$ mason external --setup
Installing Spack Registry ...
Updating Spack backend ...
fatal: tag 'v0.14.2' already exists
Spack update failed.
```

Do you see this on your end?

Everything else I tried worked great. Nice work.

ankingcodes commented:
@ben-albrecht I don't see this message on my side. I assume, the tag exists on your `spack` because of previously done tests. If you remove `spack` and clone `releases/v0..13` branch and then try the test, there shouldn' t be any error message. Another alternative could be to delete the tag and try the test again.

ben-albrecht commented:
> I don't see this message on my side. I assume, the tag exists on your spack because of previously done tests.

I'm seeing it with a fresh install of spack - here are the full steps of what I'm doing:

```sh
$ rm -rf ~/.mason/spack
$ rm -rf ~/.mason/spack-registry

# master branch
$ mason external --setup
Installing Spack backend ...

# ankingcodes:setup branch (#15641)
$ mason external --setup
Installing Spack Registry ...
Updating Spack backend ...
fatal: tag 'v0.14.2' already exists
Spack update failed.
```

The error is coming from `git -C $SPACK_ROOT tag v0.14 -m "updated"` (`gitCommitTag`), but I haven't figure out why yet.

As a sanity check, I do not have `SPACK_ROOT` set, and am using `chpl version 1.23.0 pre-release` (991c6d581b) as my master branch.

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -79,26 +113,115 @@ private proc specHelp() {
+  }
+  if getSpackVersion != spackVersion && SPACK_ROOT == spackDefaultPath {
+    throw new owned MasonError("Mason has been updated and requires a newer" +
+          " version of spack.\nTo use `mason external`, call `mason external --setup`");
```
E.g.,

```suggestion
          " version of Spack.\nTo use mason external, call: mason external --setup");
```

ankingcodes commented:
@ben-albrecht The issue was that, in an earlier version of `spack` all the tags from the repo were present including `v0.14.2`, hence we got the error message. I've applied a `-f` i.e. , `--force`  option which will update a tag even if it is  already present. But `git tag` doesn't have a `-q` or `--quiet` option due to which we get a message printed as : `Updated tag 'v0.14.2' (was 7a68a4d85) ` 
To get rid of that, I've applied a small hack ` &> /dev/null` which is making the command quiet. Let me know if it works for you. 
I didn't get any errors on my side. Check on your end once 👍

ankingcodes commented:
Any clue why 80e1650 failed ?

ben-albrecht commented:
> But git tag doesn't have a -q or --quiet option due to which we get a message printed as : Updated tag 'v0.14.2' (was 7a68a4d85) 

Remind me why we use `git tag` again? Should we be checking out the specific tag instead with `git checkout -C ... tags/v<version> -b v<version>-branch` ?

ankingcodes commented:
>  why we use git tag again?

In this case, tags are basically release points. For the earlier `spack`, we used branch `releases/v0.13` and tag `v0.13.2`. So for the latest spack, we first fetched the `releases/v0.14` branch , then we checked out the `FETCH_HEAD` to a new branch `v{version}-branch` and we updated/created tag `v0.14.2`. If a tag exists, then we overwrite it, however, this won't be the case for future releases, ie., there won't be an existing tag for `v0.15`.
The commands are as follows : 
```
// fetch
git -C SPACK_ROOT fetch -q --depth 1 origin releases/v0.14
// checkout
git -C SPACK_ROOT checkout -q FETCH_HEAD -b 0.14.2-branch
// create tag
git -C SPACK_ROOT tag v0.14.2 -f -m "updated" &> /dev/null
```
We should also make sure the `repos.yaml` file should be pointing to `spack-registry` so I am adding `generateYAML` to both updating and creating of `spack`

ankingcodes commented:
> Should we be checking out the specific tag instead with git checkout -C ... tags/v<version> -b v<version>-branch ?

I don't think so. It'd be good for now, but for future versions for which a tag is not currently present in the `spack` repo, we would require to implement a separate update function which does basically the same thing.

ben-albrecht commented:
Sorry to not bring this up earlier. Thinking about this a bit more - I wonder if we should _only_ be checking out tags (rather than any specific branch), since we want to pin the version that was released.

Can you see any issue with the following process? 

Checking out a specific version:

```sh
# Shallow clone of spack at the v0.14.2 tag
git clone -q --depth 1 --branch v0.14.2 https://github.com/spack/spack
```

Upgrading an existing checkout (in this case, downgrading to provide a working example):

```
# Fetch the tag we want to upgrade to
git fetch -q --depth 1 origin refs/tags/v0.14.1:refs/tags/v0.14.1
# Checkout the tag
git checkout -q v0.14.1
```

Also, the command `git fetch -q --depth 1 origin refs/tags/v0.14.1:refs/tags/v0.14.1` doesn't appear to generate an error if the tag already exists, which should eliminate the issue we're seeing when upgrading from an old mason version to a new mason version.

ankingcodes commented:
@ben-albrecht Okay. I am working on it. 

Just to make sure we are on the same page, if an user doesn't have spack & spack-registry installed, `setupSpack()` does the following: 
```
# clone branch releases/v{version} to SPACK_ROOT
git clone -q --branch releases/v{version} --depth 1 https://github.com/spack/spack SPACK_ROOT

# clone branch master to MASON_HOME + /spack-registry
git clone -q releases --branch master --depth 1 https://github.com/spack/spack MASON_HOME
```
Do you see any problem here ?

ben-albrecht commented:
> Do you see any problem here ?

One slight correction: we should check out a branch from the tag directly (`v0.14.2`) rather than the release branch (`releases/v0.14.2`), so I would update the first command to be:

```sh
# clone branch releases/v{version} to SPACK_ROOT
git clone -q --branch v{version} --depth 1 https://github.com/spack/spack SPACK_ROOT
```

ankingcodes commented:
@ben-albrecht I just tried that out. The problem is that we get a git error message as follows and we have a detached branch : 
```
Note: checking out '7a68a4d8513cd6105ee21bedfe04664c292d1e7d'. 

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>
```
I had used `releases/v0.14` branch to counter this error earlier.

ben-albrecht commented:
@ankingcodes - is being on a detached HEAD an issue other than the warning itself?

We can remove the warning with a git config modification on the command-line:

```sh
git clone -c advice.detachedHead=false  -q --branch v{version} --depth 1 https://github.com/spack/spack SPACK_ROOT
```

ankingcodes commented:
> is being on a detached HEAD an issue other than the warning itself?

The git documentation advices not to be on a detached HEAD. However, if we check it out to a new branch after clone, there should not be any problem.

ben-albrecht commented:
> The git documentation advices not to be on a detached HEAD

As long as we're never doing any commands in the spack repo beyond fetches and checkouts, I think it's fine. We could revisit if we ran into a case that require us to be on an actual branch rather than a detached HEAD though.

ankingcodes commented:
@ben-albrecht Okay, if an user has installed spack today, he only gets a single tag `v0.14.2` on his SPACK_ROOT. If on a next version, suppose `v0.15.0`, user would see an error message to update spack via `mason external --setup`. But when user runs it, he would get a `fatal : v0.15.0 tag not found` since we are checking out a tag that doesn't exists. How would that be handled, would there be something like `mason external --update` ?

ben-albrecht commented:
> But when user runs it, he would get a fatal : v0.15.0 tag not found since we are checking out a tag that doesn't exists. How would that be handled, would there be something like mason external --update ?

I did not see an error when running the following - perhaps this could be our logic for updating an existing spack checkout?

```sh
# Fetch the tag we want to upgrade to
git fetch -q --depth 1 origin refs/tags/v0.14.1:refs/tags/v0.14.1
# Checkout the tag
git checkout -q v0.14.1
```

ankingcodes commented:
@ben-albrecht Made the changes. Also re-checked few cases. Everything works fine !
The "detached HEAD" can be left for future work, if someone faces an issue because of it.

ankingcodes commented on tools/mason/MasonUpdate.chpl:
```diff
@@ -66,12 +66,13 @@ proc UpdateLock(args: list(string), tf="Mason.toml", lf="Mason.lock") {
     const TomlFile = parseToml(openFile);
+    if !isDir(SPACK_ROOT) || !isDir(MASON_HOME+'/spack-registry') then
+    throw new owned MasonError("Mason could not find Spack at $MASON_HOME. " +
+                  "To install Spack, use `mason external --setup`.");
```
@ben-albrecht I think we should remove this because if an user doesn't want to use spack, and doesn't have spack installed, they might also see this message. Perhaps should first check if SPACK_ROOT and spack-registry is present, and then check if version is not updated and then show the error. What do u think?

ben-albrecht replied:
I agree. I think the `MasonBuild` check when `external` is present in the manifest is sufficient. Good catch!

ankingcodes replied:
I've added a check in 48186a2. I believe that should be sufficient ?

ben-albrecht replied:
Yes, I think so. Thanks

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -41,14 +47,40 @@ proc masonExternal(args: [] string) {
+      getSpackVersion == spackVersion &&
+      SPACK_ROOT == spackDefaultPath {
         throw new owned MasonError("Spack backend is already installed");
-      else if MASON_OFFLINE then
```
Is `MASON_OFFLINE` still respected with these changes? Setting `MASON_OFFLINE` should disable `mason external --setup` still.

ankingcodes replied:
Done 👍

ben-albrecht commented on tools/mason/MasonUpdate.chpl:
```diff
@@ -66,12 +66,15 @@ proc UpdateLock(args: list(string), tf="Mason.toml", lf="Mason.lock") {
     updateRegistry(tf, args);
     const openFile = openreader(tomlPath);
     const TomlFile = parseToml(openFile);
+    if isDir(SPACK_ROOT) {
```
@ankingcodes - Does this catch any cases that are not already caught in `MasonBuild`?

If we keep this check, we need to update the error message.

When upgrading from the previous version of `mason` and doing a `mason build`, we get this error message:

```
Mason could not find Spack at $MASON_HOME.
To install Spack, use : mason external --setup
```

where the other error message is probably more fitting:

```
Mason has been updated. To use mason external, call : mason external --setup
```

ankingcodes replied:
I'm in favour of keeping this check here. I'll update the error message.

ben-albrecht replied:
OK, will this error message fire if a user is building (and updating) a project that does not use any external packages?

ankingcodes replied:
Yes, if an user has a SPACK_ROOT that hasn't been updated.

ben-albrecht replied:
I think we only want this error message to fire if the user is doing anything that would require external packages or `mason external` commands. That way, a user that never ran `mason external --setup` before will never run into any of these errors.

ankingcodes replied:
I'll add an additional check for checking 'external' in TOML file.

ben-albrecht commented on test/mason/mason-external/libtomlc99/mason-external.good:
```diff
@@ -1,4 +1,6 @@
+  Fetch: <time>.  Build: <time>.  Total: <time>.
 Installing Spack backend ...
+Spack installation successful
```
Do we still print this message? When running this test locally, I got an error because this line was missing.

ankingcodes replied:
I think you had mentioned in a previous comment that nothing else should be printed other than error messages during spack installation, so I removed that. Would you want that to be added back?

ben-albrecht replied:
Gotcha - I think it's fine to leave it out. Let's just update this `.good` :)

ankingcodes replied:
Also, should I get rid of the `SKIPIF` from the test now ?

ben-albrecht replied:
Let's have me do that in a follow-up PR after this is merged. I will need to test manually on several of our systems before we do that to catch any unexpected issues.

ankingcodes replied:
Cool !

ankingcodes commented:
@ben-albrecht changes done. I think its good for merge now.

ben-albrecht commented on tools/mason/MasonUpdate.chpl:
```diff
@@ -66,12 +66,15 @@ proc UpdateLock(args: list(string), tf="Mason.toml", lf="Mason.lock") {
+    if isDir(SPACK_ROOT) && TomlFile.pathExists('external') {
+      if getSpackVersion != spackVersion then
+      throw new owned MasonError("Mason has been updated. " +
+                  "To install Spack, call  : mason external --setup.");
```
```suggestion
                  "To install Spack, call: mason external --setup.");
```

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -79,28 +118,108 @@ private proc specHelp() {
+  }
+  if getSpackVersion != spackVersion && SPACK_ROOT == spackDefaultPath {
+    throw new owned MasonError("Mason has been updated and requires a newer" +
+          " version of Spack.\nTo use mason external, call : mason external --setup");
```
```suggestion
          " version of Spack.\nTo use mason external, call: mason external --setup");
```

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -79,28 +118,108 @@ private proc specHelp() {
+  }
+  if !isDir(getSpackRegistry) {
+    throw new owned MasonError("Mason has been updated. To use mason external, "+
+                                "call : mason external --setup");
```
```suggestion
                                "call: mason external --setup");
```

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -79,28 +118,108 @@ private proc specHelp() {
+proc spackInstalled() throws {
   if !isDir(SPACK_ROOT) {
-    throw new owned MasonError("To use `mason external` call `mason external --setup`");
+    throw new owned MasonError("To use mason external, call : mason external --setup");
```
```suggestion
    throw new owned MasonError("To use mason external, call: mason external --setup");
```

ankingcodes commented:
@ben-albrecht Made the changes.
> FWIW, it installs in 9 seconds for me now

WOW 🎉 💯

ben-albrecht commented on tools/mason/MasonExternal.chpl:
```diff
@@ -79,28 +118,108 @@ private proc specHelp() {
+    remove(yamlFilePath);
+  }
+  const reposOverride = 'repos:\n'+
+                        '  - ~/.mason/spack-registry/var/spack/repos/builtin \n';
```
@ankingcodes - in doing some testing for removing the SKIPIF, I ran into an issue here -- this should be `MASON_HOME` instead of `~/.mason`.

ankingcodes replied:
@ben-albrecht  Okay, I'll make a PR for this.

ben-albrecht replied:
Thanks!
